### PR TITLE
new: Added jina clip v1

### PR DESCRIPTION
**Pull request by hh-space-invader**
Adding jinaai/jina-clip-v1
They provided two examples, the first one works and the second one complains about missing `jinaai/jina-clip-v1/sentence_xlnet_config.json`. The output of the first one seems to have small numbers, like they are normallized but its not mentioned so not sure tbh.

Update:
The text model needs pooling and normalizing
The image model needs the image to be square

### All Submissions:

* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

* [x] Does your submission pass the existing tests?
* [x] Have you added tests for your feature?
* [x] Have you installed `pre-commit` with `pip3 install pre-commit` and set up hooks with `pre-commit install`?

### New models submission:

* [ ] Have you added an explanation of why it's important to include this model?
* [ ] Have you added tests for the new model? Were canonical values for tests computed via the original model?
* [ ] Have you added the code snippet for how canonical values were computed?
* [ ] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: fastembed/image/transform/functional.py [hunk 1]
```diff
@@ -122,3 +122,18 @@ def pil2ndarray(image: Union[Image.Image, np.ndarray]):
...
+    fill_color: Optional[Union[str, int, tuple[int, ...]]] = None,
+    resample: Union[Image.Resampling, int] = Image.Resampling.BICUBIC,
+) -> Image.Image:
+    resized_image = resize(image=image, size=size, resample=resample)
+
+    new_image = Image.new(mode="RGB", size=(size, size), color=fill_color)
+    left = (size - resized_image.size[0]) // 2
+    top = (size - resized_image.size[1]) // 2
+    new_image.paste(resized_image, (left, top))
+    return new_image
```
Comment (joein) [inline fastembed/image/transform/functional.py:139]: we already have `resize`, let's not introduce new functions, we can add an additional parameter like `preserve_aspect_ratio` or something like this to `resize`.
If it's false then we should just resize image to the required size (preserving aspect ratio is useful when later we have crop)

File: fastembed/image/transform/operators.py [hunk 2]
```diff
@@ -188,5 +224,50 @@ def _get_rescale(transforms: list[Transform], config: dict[str, Any]):
...
     def _get_normalize(transforms: list[Transform], config: dict[str, Any]):
-        if config.get("do_normalize", False):
-            transforms.append(Normalize(mean=config["image_mean"], std=config["image_std"]))
+        if config.get("do_normalize", False) or ("mean" in config and "std" in config):
+            transforms.append(
+                Normalize(
+                    mean=config.get("image_mean", config.get("mean")),
+                    std=config.get("image_std", config.get("std")),
+                )
+            )
```
Comment (joein) [inline fastembed/image/transform/operators.py:233]:
```suggestion
if config.get("do_normalize", False):
    transforms.append(Normalize(mean=config["image_mean"], std=config["image_std"]))
elif "mean" in config and "std" in config:
    transforms.append(Normalize(mean=config["mean"], std=config["std"]))
```

**Review by joein — CHANGES_REQUESTED** [review]

File: fastembed/image/transform/operators.py [hunk 3]
```diff
@@ -157,6 +182,23 @@ def _get_resize(transforms: list[Transform], config: dict[str, Any]):
                         resample=config.get("resample", Image.Resampling.BICUBIC),
                     )
                 )
+        elif mode == "JinaCLIPImageProcessor":
+            resample = (
+                Compose._interpolation_resolver(config.get("interpolation"))
+                if isinstance(config.get("interpolation"), str)
+                else config.get("interpolation") or Image.Resampling.BICUBIC
```
Comment (I8dNLo) [inline fastembed/image/transform/operators.py:189]: Could we just put Image.Resampling.BICUBIC in a get block?

File: fastembed/image/transform/operators.py [hunk 4]
```diff
@@ -173,6 +215,8 @@ def _get_center_crop(transforms: list[Transform], config: dict[str, Any]):
                 transforms.append(CenterCrop(size=crop_size))
         elif mode == "ConvNextFeatureExtractor":
             pass
+        elif mode == "JinaCLIPImageProcessor":
```
Comment (I8dNLo) [inline fastembed/image/transform/operators.py:218]: Could reorganize it to 'if mode in ...'

File: fastembed/image/transform/operators.py [hunk 5]
```diff
@@ -190,3 +234,36 @@ def _get_rescale(transforms: list[Transform], config: dict[str, Any]):
...
+        elif "mean" in config and "std" in config:
+            transforms.append(Normalize(mean=config["mean"], std=config["std"]))
+
+    @staticmethod
+    def _get_pad2square(transforms: list[Transform], config: dict[str, Any]):
+        mode = config.get("image_processor_type", "CLIPImageProcessor")
+        if mode == "CLIPImageProcessor":
+            pass
+        elif mode == "ConvNextFeatureExtractor":
+            pass
```
Comment (I8dNLo) [inline fastembed/image/transform/operators.py:246]: Also here

**Review by I8dNLo — APPROVED** [review]

File: fastembed/image/transform/functional.py [hunk 6]
```diff
@@ -122,3 +122,26 @@ def pil2ndarray(image: Union[Image.Image, np.ndarray]):
...
+
+def pad2square(
+    image: Image.Image,
+    size: int,
+    fill_color: Optional[Union[str, int, tuple[int, ...]]] = None,
+) -> Image.Image:
+    height, width = image.height, image.width
+
+    # if the size is larger than the new canvas
+    if width > size or height > size:
```
Comment (joein) [inline fastembed/image/transform/functional.py:135]: should not it be `if width >= size and height >= size` ?

File: fastembed/image/transform/functional.py [hunk 7]
```diff
@@ -122,3 +122,26 @@ def pil2ndarray(image: Union[Image.Image, np.ndarray]):
     if isinstance(image, Image.Image):
         return np.asarray(image).transpose((2, 0, 1))
     return image
+
+
+def pad2square(
+    image: Image.Image,
+    size: int,
+    fill_color: Optional[Union[str, int, tuple[int, ...]]] = None,
```
Comment (joein) [inline fastembed/image/transform/functional.py:130]: what happens if `fill_color` is None ?

File: fastembed/image/transform/functional.py [hunk 8]
```diff
@@ -122,3 +122,26 @@ def pil2ndarray(image: Union[Image.Image, np.ndarray]):
...
+    # if the size is larger than the new canvas
+    if width > size or height > size:
+        left = (width - size) // 2
+        top = (height - size) // 2
+        right = left + size
+        bottom = top + size
+        image = image.crop((left, top, right, bottom))
+        return image
+
+    new_image = Image.new(mode="RGB", size=(size, size), color=fill_color)
```
Comment (joein) [inline fastembed/image/transform/functional.py:143]: What if we pass a grayscale image?

File: fastembed/image/transform/operators.py [hunk 9]
```diff
@@ -37,7 +38,9 @@ def __init__(self, mean: Union[float, list[float]], std: Union[float, list[float
         self.std = std
 
     def __call__(self, images: list[np.ndarray]) -> list[np.ndarray]:
-        return [normalize(image, mean=self.mean, std=self.std) for image in images]
+        return [
+            normalize(image, mean=np.array(self.mean), std=np.array(self.std)) for image in images
```
Comment (joein) [inline fastembed/image/transform/operators.py:42]: why? normalize accepts float as well

File: fastembed/image/transform/operators.py [hunk 10]
```diff
@@ -157,6 +182,23 @@ def _get_resize(transforms: list[Transform], config: dict[str, Any]):
                         resample=config.get("resample", Image.Resampling.BICUBIC),
                     )
                 )
+        elif mode == "JinaCLIPImageProcessor":
+            resample = (
+                Compose._interpolation_resolver(config.get("interpolation"))
+                if isinstance(config.get("interpolation"), str)
+                else config.get("interpolation", Image.Resampling.BICUBIC)
+            )
```
Comment (joein) [inline fastembed/image/transform/operators.py:190]: it is hard to read, why do we do `config.get('interpolation')` 3 times? We cna just write it to a variable once and reuse

File: fastembed/image/transform/operators.py [hunk 11]
```diff
@@ -190,3 +234,36 @@ def _get_rescale(transforms: list[Transform], config: dict[str, Any]):
...
+            "bilinear": Image.Resampling.BILINEAR,
+            "bicubic": Image.Resampling.BICUBIC,
+            "box": Image.Resampling.BOX,
+            "hamming": Image.Resampling.HAMMING,
+        }
+
+        if resample and (method := interpolation_map.get(resample.lower())):
+            return method
+
+        raise ValueError(f"Unknown interpolation method: {resample}")
```
Comment (joein) [inline fastembed/image/transform/operators.py:269]: feels like it should not be a part of `Compose` class

File: fastembed/text/pooled_embedding.py [hunk 12]
```diff
@@ -50,6 +50,10 @@ def _get_worker_class(cls) -> Type[TextEmbeddingWorker]:
 
     @classmethod
     def mean_pooling(cls, model_output: np.ndarray, attention_mask: np.ndarray) -> np.ndarray:
+        if model_output.ndim == 2:  # (batch, embedding_dim)
+            seq_length = attention_mask.shape[1]
+            # (batch, seq_length, embedding_dim)
+            model_output = np.tile(np.expand_dims(model_output, axis=1), (1, seq_length, 1))
```
Comment (joein) [inline fastembed/text/pooled_embedding.py:56]: is it actually required? how did it work before?

**Review by joein — CHANGES_REQUESTED** [review]

File: fastembed/image/transform/operators.py (hunk 11 above)
Comment (joein) [reply fastembed/image/transform/operators.py:268]: idk for sure, we can just move it out of the class
at least, this `Compose._interpolation_resolver` is super ugly, if we keep this method here, we need to make `get_resize` a class method, not a static, it is only used inside of `Compose` class anyway

File: fastembed/text/pooled_embedding.py (hunk 12 above)
Comment (joein) [reply fastembed/text/pooled_embedding.py:56]: Mean pooling is applied to token embeddings, however, it seems that jina provides sentence embeddings straightaway, so pooling is not required.
Also I don't really get the point about 1D and 2D
1) I don't understand why images are related to the text embedding class
2) As it have mentioned, since mean pooling is applied to token embeddings, expected shape is (batch_size, num_tokens, dim)

**Review by joein — CHANGES_REQUESTED** [review]
please look at the comments above

**Review by joein — APPROVED** [review]
